x/eibc/keeper/keeper.go

```diff
@@ -0,0 +1,178 @@
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	commontypes "github.com/dymensionxyz/dymension/v3/x/common/types"
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/dymensionxyz/dymension/v3/x/eibc/types"
+)
+
+type (
+	Keeper struct {
+		cdc        codec.BinaryCodec
+		storeKey   storetypes.StoreKey
+		memKey     storetypes.StoreKey
+		hooks      types.EIBCHooks
+		paramstore paramtypes.Subspace
+		types.AccountKeeper
+		types.BankKeeper
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey,
+	memKey storetypes.StoreKey,
+	ps paramtypes.Subspace,
+	accountKeeper types.AccountKeeper,
+	bankKeeper types.BankKeeper,
+
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		cdc:           cdc,
+		storeKey:      storeKey,
+		memKey:        memKey,
+		paramstore:    ps,
+		AccountKeeper: accountKeeper,
+		BankKeeper:    bankKeeper,
+	}
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) SetDemandOrder(ctx sdk.Context, order *types.DemandOrder) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.DemandOrderKeyPrefix))
+	//TODO: FIXME -  Need to write a method of update demand order with status
+	store.Set(types.GetDemandOrderKey(order.TrackingPacketStatus.String(), order.Id), k.cdc.MustMarshal(order))
+
+	// Emit events
+	eventAttributes := []sdk.Attribute{
+		sdk.NewAttribute(types.AttributeKeyPacketKey, order.Id),
+		sdk.NewAttribute(types.AttributeKeyPrice, order.Price.String()),
+		sdk.NewAttribute(types.AttributeKeyFee, order.Fee.String()),
+		sdk.NewAttribute(types.AttributeKeyIsFullfilled, strconv.FormatBool(order.IsFullfilled)),
+		sdk.NewAttribute(types.AttributeKeyPacketStatus, order.TrackingPacketStatus.String()),
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeEIBC,
+			eventAttributes...,
+		),
+	)
+
+}
+
+func (k Keeper) deleteDemandOrder(ctx sdk.Context, order *types.DemandOrder) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.DemandOrderKeyPrefix))
+	store.Delete(types.GetDemandOrderKey(order.TrackingPacketStatus.String(), order.Id))
+}
+
+// UpdateDemandOrderWithStatus deletes the current demand order and creates a new one with and updated packet status under a new key.
+// Updating the status should be called only with this method as it effects the key of the packet.
+// The assumption is that the passed demand order packet status field is not updated directly.
+func (k *Keeper) UpdateDemandOrderWithStatus(ctx sdk.Context, demandOrder *types.DemandOrder, newStatus commontypes.Status) *types.DemandOrder {
```

Comment:
when it's used?
as I see it's called only when changed from PENDING to FINALIZED
and not sure why we need to update the DemandOrder.